src/libertem/executor/pipelined.py

```diff
@@ -0,0 +1,857 @@
+        self._startup_timeout = startup_timeout
+
+        # keep this at the bottom:
+        self._pool = self._start_pool()
+
+    def _start_pool(self) -> WorkerPool:
+        with tracer.start_as_current_span("PipelinedExecutor.start_pool") as span:
+            pool = WorkerPool(
+                worker_fn=functools.partial(
+                    pipelined_worker,
+                    pin=self._pin_workers,
+                    early_setup=self._early_setup,
+                ),
+                spec=self._spec,
+            )
+            # FIXME: check process liveness here, too, to catch early startup
+            # failures without having to run into the timeout
+            for i in range(pool.size):
+                with pool.response_queue.get(timeout=self._startup_timeout) as (msg, _):
+                    if msg["type"] == "ERROR":
+                        raise RuntimeError(
+                            f"error on startup: {msg['error']}"
+                        )
+                    if msg["type"] != "STARTUP_DONE":
+                        raise RuntimeError(
+                            f"unknown message type {msg['type']}, expected STARTUP_DONE"
+                        )
+                    span.add_event("worker startup done", {"worker_id": msg["worker_id"]})
+            # set here, so we don't try to close the pool if it doesn't exist
+            self._closed = False
+            return pool
+
+    @classmethod
+    def _default_spec(cls):
+        from libertem.utils.devices import detect
+        detected = detect()
+        return _make_spec(**detected)
+
+    @classmethod
+    def make_local(cls, **kwargs):
+        """
+        Create a :code:`PipelinedExecutor` with the default spec.
+        """
+        spec = cls._default_spec()
+        return cls(spec=spec, **kwargs)
+
+    @classmethod
+    def make_spec(cls, *args, **kwargs):
+        return _make_spec(*args, **kwargs)
+    make_spec.__doc__ = _make_spec.__doc__
+
+    def _validate_worker_state(self):
+        if not self._pool.all_alive():
+            raise RuntimeError("some workers are stopped, cannot continue")
+
+    def _run_tasks_inner(
+        self,
+        tasks: Iterable[TaskProtocol],
+        params_handle: Any,
+        cancel_id: Any,
+        task_comm_handler: "TaskCommHandler",
+    ) -> ResultWithID:
+        in_flight = 0
+        id_to_task = {}
+
+        try:
+            self._validate_worker_state()
+            task_comm_handler.start()
+            span = trace.get_current_span()
+            span_context = span.get_span_context()
+
+            for qs, _ in self._pool.workers:
+                qs.request.put({
+                    "type": "WARMUP",
+                    "span_context": span_context,
+                })
+
+            for task_idx, task in enumerate(tasks):
+                in_flight += 1
+                id_to_task[task_idx] = task
+
+                worker_idx = task_idx % self._pool.size
```

Comment:
I think both the threading and the mulitprocessing queue implement `qsize()` to get the approximate queue size - so it's not guaranteed to be correct, but for this purpose that should be enough. I did not check overhead.

For the first version I'd like to keep the simple "static" scheduling, but I'll add a note in the code to indicate that it can be improved. Maybe something for 0.11?

Reply:
Agreed :+1: